src/main/java/org/broadinstitute/hellbender/tools/LocalAssembler.java

```diff
@@ -0,0 +1,2306 @@
+        @Override public boolean isGap() { return true; }
+        @Override public int getLength() { return sequence.length() - Kmer.KSIZE + 1; }
+        @Override public PathPart rc() { return new PathPartGap(new SequenceRC(sequence)); }
+        @Override public char getFirstCall() { return sequence.charAt(Kmer.KSIZE - 1); }
+        @Override public char getLastCall() {
+            return sequence.charAt(sequence.length() - Kmer.KSIZE + 1);
+        }
+    }
+
+    /** A part of a path that is present as a sub-sequence of some Contig. **/
+    public static final class PathPartContig implements PathPart {
+        private final Contig contig;
+        private final int start;
+        private int stop;
+
+        public PathPartContig( final Contig contig, final int start ) {
+            this(contig, start, start+1);
+        }
+        public PathPartContig( final Contig contig, final int start, final int stop ) {
+            this.contig = contig;
+            this.start = start;
+            this.stop = stop;
+        }
+
+        @Override public Contig getContig() { return contig; }
+        @Override public String getSequence() { return null; }
+        @Override public void extend( final char call ) { stop += 1; }
+        @Override public int getStart() { return start; }
+        @Override public int getStop() { return stop; }
+        @Override public boolean isGap() { return false; }
+        @Override public int getLength() { return stop - start; }
+        @Override public PathPart rc() {
+            final int revBase = contig.size() - Kmer.KSIZE + 1;
+            return new PathPartContig(contig.rc(), revBase - stop, revBase - start);
+        }
+        @Override public char getFirstCall() {
+            return getContig().getSequence().charAt(start + Kmer.KSIZE - 1);
+        }
+        @Override public char getLastCall() { return getContig().getSequence().charAt(stop - 1); }
+    }
+
+    /** A path through the assembly graph for something (probably a read). **/
+    public static final class Path {
+        private final List<PathPart> parts;
+
+        public Path( final byte[] calls,
+                     final KmerSet<KmerAdjacency> kmerAdjacencySet ) {
+            parts = new ArrayList<>();
+            long kVal = 0;
+            int count = 0;
+            PathPart currentPathPart = null;
+            for ( int idx = 0; idx != calls.length; ++idx ) {
+                final char call = (char)calls[idx];
+                kVal <<= 2;
+                switch ( call ) {
+                    case 'C': case 'c': kVal += 1; break;
+                    case 'G': case 'g': kVal += 2; break;
+                    case 'T': case 't': kVal += 3; break;
+                }
+                if ( ++count >= Kmer.KSIZE ) {
+                    final KmerAdjacency kmer = KmerAdjacencyImpl.find(kVal, kmerAdjacencySet);
+                    // if we fail to look up the kmer
+                    if ( kmer == null ) {
```

Comment:
The nested `if`s get pretty hairy in this method. Any chance you could pull some of the clauses into little methods for readability. For example this could be:

`if (kmer == null) {
  currentPathPart = addGapKmer(...)
} else {
 ...
}`

I just find it hard to match up so many levels of indentation in this method.